packages/react-next/src/components/ComboBox/ComboBox.tsx

```diff
@@ -104,34 +97,114 @@ interface IComboBoxOptionWrapperProps extends IComboBoxOption {
 }
 
 /**
- * Internal class that is used to wrap all ComboBox options.
+ * Internal component that is used to wrap all ComboBox options.
  * This is used to customize when we want to rerender components,
  * so we don't rerender every option every time render is executed.
  */
-class ComboBoxOptionWrapper extends React.Component<IComboBoxOptionWrapperProps, {}> {
-  public render(): React.ReactNode {
-    return this.props.render();
-  }
-
-  public shouldComponentUpdate(newProps: IComboBoxOptionWrapperProps): boolean {
+const ComboBoxOptionWrapper = React.memo(
+  ({ render }: IComboBoxOptionWrapperProps) => render(),
+  (
+    { render: oldRender, ...oldProps }: IComboBoxOptionWrapperProps,
+    { render: newRender, ...newProps }: IComboBoxOptionWrapperProps,
+  ) =>
     // The render function will always be different, so we ignore that prop
-    return !shallowCompare({ ...this.props, render: undefined }, { ...newProps, render: undefined });
-  }
-}
+    shallowCompare(oldProps, newProps),
+);
 
 const COMPONENT_NAME = 'ComboBox';
+const DEFAULT_PROPS: Partial<IComboBoxProps> = {
+  options: [],
+  allowFreeform: false,
+  autoComplete: 'on',
+  buttonIconProps: { iconName: 'ChevronDown' },
+};
+
+function useOptionsState({ options, defaultSelectedKey, selectedKey }: IComboBoxProps) {
+  /** The currently selected indices */
+  const [selectedIndices, setSelectedIndices] = React.useState<number[]>(() =>
+    getSelectedIndices(options, buildDefaultSelectedKeys(defaultSelectedKey, selectedKey)),
+  );
+  /** The options currently available for the callout */
+  const [currentOptions, setCurrentOptions] = React.useState<IComboBoxOption[]>(options);
+  /** This value is used for the autocomplete hint value */
+  const [suggestedDisplayValue, setSuggestedDisplayValue] = React.useState<string>();
 
-@customizable('ComboBox', ['theme', 'styles'], true)
-export class ComboBox extends React.Component<IComboBoxProps, IComboBoxState> {
-  public static defaultProps: IComboBoxProps = {
-    options: [],
-    allowFreeform: false,
-    autoComplete: 'on',
-    buttonIconProps: { iconName: 'ChevronDown' },
-  };
+  React.useEffect(() => {
+    if (defaultSelectedKey === undefined) {
+      const selectedKeys: string[] | number[] = buildSelectedKeys(selectedKey);
+      const indices: number[] = getSelectedIndices(options, selectedKeys);
 
-  private _root = React.createRef<HTMLDivElement>();
+      setSelectedIndices(indices);
+    }
+    setCurrentOptions(options);
+  }, [defaultSelectedKey, options, selectedKey]);
+
+  React.useEffect(() => {
+    if (selectedKey === null) {
+      setSuggestedDisplayValue(undefined);
+    }
+  }, [selectedKey]);
```

Comment:
To replicate the previous conditions where this was called, I think you'd need to add `text` and `options` to the dependencies. Though I'm not sure if it will matter.